#### TROUVA 

In [1]:
import json
import pandas as pd

In [2]:
data = pd.read_json('sizes.json', lines=True)


In [3]:
data.head()

__v                                   _id  \
0    0  {'$oid': '557abe8565950003006fefe7'}   
1    0  {'$oid': '557abe8565950003006fefe5'}   
2    0  {'$oid': '557abe8565950003006fefe8'}   
3    0  {'$oid': '557abe8565950003006fefe6'}   
4    0  {'$oid': '557abe8565950003006fefe9'}   

                              created_at filter_label label  old_label  \
0  {'$date': '2015-06-12T11:12:05.248Z'}            S    26      [26\]   
1  {'$date': '2015-06-12T11:12:05.244Z'}           XS    24  [24, 24\]   
2  {'$date': '2015-06-12T11:12:05.249Z'}            S    28      [28\]   
3  {'$date': '2015-06-12T11:12:05.247Z'}            S    25       [25]   
4  {'$date': '2015-06-12T11:12:05.249Z'}            M    30      [30\]   

                              updated_at  
0  {'$date': '2015-07-07T16:29:11.868Z'}  
1  {'$date': '2015-07-07T16:27:03.743Z'}  
2  {'$date': '2015-07-07T16:36:22.139Z'}  
3  {'$date': '2015-07-07T16:29:03.370Z'}  
4  {'$date': '2015-07-07T16:44:25.382Z'}

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51210 entries, 0 to 51209
Data columns (total 7 columns):
__v             51210 non-null int64
_id             51210 non-null object
created_at      51210 non-null object
filter_label    294 non-null object
label           51210 non-null object
old_label       51210 non-null object
updated_at      51210 non-null object
dtypes: int64(1), object(6)
memory usage: 2.7+ MB


**Observations**
- dataframe has 7 columns/ 51210 lines
- filter_label column has only 294 entries, vs 51 210 lines
- dataframe entries are objects which is not the most suitable type for data manipulation

**Data frame columns - detailed view**

In [5]:
column_names=list(data)

In [6]:
column_types=[type(data._id[0]),
type(data.__v[0]),
type(data.created_at[0]),
type(data.filter_label[0]),
type(data.label[0]),
type(data.old_label[0]),
type(data.updated_at[0])]

In [7]:
for i in range (0,len(column_names)):
    print(column_names[i],column_types[i])

__v <class 'dict'>
_id <class 'numpy.int64'>
created_at <class 'dict'>
filter_label <class 'str'>
label <class 'str'>
old_label <class 'list'>
updated_at <class 'dict'>


** Step 1 - flatten the dataframe** 

1.1 Column **_if**: explore dictionary keys and values

In [8]:
keys=[]
values=[]
for index,row in data.iterrows():
    item=row['_id']
    keys.extend(item.keys())
    values.extend(item.values())


In [9]:
unq_k=set(keys)

In [10]:
len(unq_k)

1

In [11]:
unq_v=set(values)

In [12]:
len(unq_v)

51210

In [202]:
for i in range (len(data)):
    item1=data.iloc[0:,1]
    break
    

In [205]:
type(item1)

pandas.core.series.Series

**Observations**
- The keys are the same across all lines, therefore, can be removed, and only ID value can remain as item ID identifier

In [19]:
for index,row in data.iterrows():
    item=row['_id']
    value=item.values()
    print(type(value))
    data.loc[data.index[index], '_id']=value
    break


AttributeError: 'dict_values' object has no attribute 'values'

In [14]:
data.head()

__v                                   _id  \
0    0            (557abe8565950003006fefe7)   
1    0  {'$oid': '557abe8565950003006fefe5'}   
2    0  {'$oid': '557abe8565950003006fefe8'}   
3    0  {'$oid': '557abe8565950003006fefe6'}   
4    0  {'$oid': '557abe8565950003006fefe9'}   

                              created_at filter_label label  old_label  \
0  {'$date': '2015-06-12T11:12:05.248Z'}            S    26      [26\]   
1  {'$date': '2015-06-12T11:12:05.244Z'}           XS    24  [24, 24\]   
2  {'$date': '2015-06-12T11:12:05.249Z'}            S    28      [28\]   
3  {'$date': '2015-06-12T11:12:05.247Z'}            S    25       [25]   
4  {'$date': '2015-06-12T11:12:05.249Z'}            M    30      [30\]   

                              updated_at  
0  {'$date': '2015-07-07T16:29:11.868Z'}  
1  {'$date': '2015-07-07T16:27:03.743Z'}  
2  {'$date': '2015-07-07T16:36:22.139Z'}  
3  {'$date': '2015-07-07T16:29:03.370Z'}  
4  {'$date': '2015-07-07T16:44:25.382Z'}

In [18]:
type(data.iloc[0,1])

dict_values